File: Data Structures & Algorithms/valid-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def valid(node: TreeNode, lower_bound: float, upper_bound: float):
            if not node:
                return True
            elif not (lower_bound < node.val < upper_bound):
                return False
            return valid(node.left, lower_bound, node.val) and valid(node.right, node.val, upper_bound)
        return valid(root, float('-infinity'), float('infinity'))

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode deleteNode(TreeNode root, int key) {
        root = dfs(root, key);
        return root;
    }

    private TreeNode dfs(TreeNode node, int key) {
        if (node == null) {
            return null;
        }
        
        if (node.val < key) {
            node.right = dfs(node.right, key);
        } else if (node.val > key) {
            node.left = dfs(node.left, key);
        } else {
            if (node.right == null) return node.left;
            if (node.left == null) return node.right;
            TreeNode curr = node.right;
            while (curr.left != null) {
                curr = curr.left;
            }
            curr.left = node.left;
            node = node.right;
        }
        return node;
    }
}